x/emissions/module/rewards/scores.go

```diff
@@ -0,0 +1,144 @@
+package rewards
+
+import (
+	"github.com/allora-network/allora-chain/x/emissions/keeper"
+	"github.com/allora-network/allora-chain/x/emissions/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+/*
+ These functions will be used immediately after the network loss for the relevant time step has been generated.
+ Using the network loss and the sets of losses reported by each repeater, the scores are calculated. In the case
+ of workers (who perform the forecast task and network task), the last 10 previous scores will also be taken into
+ consideration to generate the score at the most recent time step.
+*/
+
+// GenerateReputerScores calculates and persists scores for reputers based on their reported losses.
+func GenerateReputerScores(ctx sdk.Context, keeper keeper.Keeper, topicId uint64, block int64, reportedLosses types.ReputerValueBundles) ([]types.Score, error) {
+	// Get reputers data
+	var reputerAddresses []sdk.AccAddress
+	var reputerStakes []float64
+	var reputerListeningCoefficients []float64
+	var losses [][]float64
+	for _, reportedLoss := range reportedLosses.ReputerValueBundles {
+		reputerAddr, err := sdk.AccAddressFromBech32(reportedLoss.Reputer)
+		if err != nil {
+			return []types.Score{}, err
+		}
+		reputerAddresses = append(reputerAddresses, reputerAddr)
+
+		// Get reputer topic stake
+		reputerStake, err := keeper.GetStakeOnTopicFromReputer(ctx, topicId, reputerAddr)
+		if err != nil {
+			return []types.Score{}, err
+		}
+		reputerStakes = append(reputerStakes, float64(reputerStake.BigInt().Int64()))
+
+		// Get reputer listening coefficient
+		res, err := keeper.GetListeningCoefficient(ctx, topicId, reputerAddr)
+		if err != nil {
+			return []types.Score{}, err
+		}
+		reputerListeningCoefficients = append(reputerListeningCoefficients, res.Coefficient)
+
+		// Get all reported losses from bundle
+		reputerLosses := ExtractValues(reportedLoss.ValueBundle)
```

Comment:
generally i think you're making it harder for yourself to debug when you index by "index" as opposed to reputer address in this case

also looking at `GetStakeWeightedLossMatrix()`, which is downstream of `GetAllReputersOutput` invoked here... i see you're not dealing with the case where reputers report on a different set of workers with potentially a different distribution regarding "who sent inference and/or forecast"... right now you're assuming they're all reporting on the same thing...

surely our own reputers will eb able to deal with this and always report on a uniform set, but we're just opening the door for malicious actors if we're not taking that into account -- cuz obvi we won't be the only people running reputers